File: 2217-step-by-step-directions-from-a-binary-tree-node-to-another/2217-step-by-step-directions-from-a-binary-tree-node-to-another.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public String getDirections(TreeNode root, int startValue, int destValue) {
        Map<Integer, TreeNode> parentMap = new HashMap<>();
        TreeNode startNode = findStartNode(root, startValue);
        populateParentMap(root, parentMap);

        Queue<PathElement> queue = new LinkedList<>();
        queue.add(new PathElement(startNode, ""));
        Set<TreeNode> visitedNodes = new HashSet<>();
        visitedNodes.add(startNode);

        while (!queue.isEmpty()) {
            PathElement currentElement = queue.poll();

            if (currentElement.node.val == destValue) {
                return currentElement.path;
            }

            if (parentMap.containsKey(currentElement.node.val)) {
                TreeNode parentNode = parentMap.get(currentElement.node.val);
                if (!visitedNodes.contains(parentNode)) {
                    queue.add(new PathElement(parentNode, currentElement.path + "U"));
                    visitedNodes.add(parentNode);
                }
            }

            if (currentElement.node.left != null && !visitedNodes.contains(currentElement.node.left)) {
                queue.add(new PathElement(currentElement.node.left, currentElement.path + "L"));
                visitedNodes.add(currentElement.node.left);
            }

            if (currentElement.node.right != null && !visitedNodes.contains(currentElement.node.right)) {
                queue.add(new PathElement(currentElement.node.right, currentElement.path + "R"));
                visitedNodes.add(currentElement.node.right);
            }
        }

        return "";
    }

    private void populateParentMap(TreeNode node, Map<Integer, TreeNode> parentMap) {
        if (node == null) return;

        if (node.left != null) {
            parentMap.put(node.left.val, node);
            populateParentMap(node.left, parentMap);
        }

        if (node.right != null) {
            parentMap.put(node.right.val, node);
            populateParentMap(node.right, parentMap);
        }
    }

    private TreeNode findStartNode(TreeNode node, int startValue) {
        if (node == null) return null;
        if (node.val == startValue) return node;

        TreeNode leftResult = findStartNode(node.left, startValue);
        if (leftResult != null) return leftResult;

        return findStartNode(node.right, startValue);
    }

    private class PathElement {
        String path;
        TreeNode node;
        PathElement(TreeNode node, String path) {
            this.node = node;
            this.path = path;
        }
    }
}

File: 2217-step-by-step-directions-from-a-binary-tree-node-to-another/2217-step-by-step-directions-from-a-binary-tree-node-to-another.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getDirections(
        self, root: TreeNode, startValue: int, destValue: int
    ) -> str:
        start_path = []
        dest_path = []

        # Find paths from root to start and destination nodes
        self._find_path(root, startValue, start_path)
        self._find_path(root, destValue, dest_path)

        directions = []
        common_path_length = 0

        # Find the length of the common path
        while (
            common_path_length < len(start_path)
            and common_path_length < len(dest_path)
            and start_path[common_path_length] == dest_path[common_path_length]
        ):
            common_path_length += 1

        # Add "U" for each step to go up from start to common ancestor
        directions.extend("U" * (len(start_path) - common_path_length))

        # Add directions from common ancestor to destination
        directions.extend(dest_path[common_path_length:])

        return "".join(directions)

    def _find_path(self, node: TreeNode, target: int, path: List[str]) -> bool:
        if node is None:
            return False

        if node.val == target:
            return True

        # Try left subtree
        path.append("L")
        if self._find_path(node.left, target, path):
            return True
        path.pop()  # Remove last character

        # Try right subtree
        path.append("R")
        if self._find_path(node.right, target, path):
            return True
        path.pop()  # Remove last character

        return False